### ISSUE TITLE
Lint: `manual_open_range_pattern`

### ISSUE THREAD

**Issue by mikerite** [rust-lang/rust-clippy#7662]
### What it does

Suggest using open range patterns instead of inclusive range patterns for integer types with upper bound equal to the maximum of that type.

### Categories

- Kind: Style

### Advantages

The suggested code is shorter and more readable. The shorter syntax doesn't require the MAX constant and so the variables type doesn't need to be repeated.

### Drawbacks

None.

Explicitly including the upper bound gives no additional information. It just repeats the type and confirms that it has an upper bound. I expect this lint to be uncontroversial.

### Example

```rust
match x as u32 {
      0 => println!("zero!"),
      1..=u32::MAX => println!("positive number!"),
}
```
Could be written as:

```rust
match x as u32 {
      0 => println!("zero!"),
      1.. => println!("positive number!"),
}
```

### Configuration

MSRV 1.55

**Comment by camsteffen** [comment]
I would call it `range_to_max` so that it makes sense with "allow".

**Comment by workingjubilee** [comment]
I made the case for stabilizing this pattern form. Based on what I remember of the code, there are some potential false positives or unwanted feedback situations to be careful of:
- Slice item patterns.
- `u16::MAX as usize` kinds of expressions.
- Other types with a `::MAX` constant, where it would not be so clear.
- Ambiguous numerics in general... remember that the presence of e.g. `i64::MAX` in code can affect type inference for surrounding numbers (though I don't think that it works the same way in patterns).
